sdlib/d/gc/extent.d

```diff
@@ -266,7 +293,59 @@ public:
 		// FIXME: in contract.
 		assert(isSlab(), "slabData accessed on non slab!");
 
-		return _metadata.slabData;
+		return _metadata.slabData.slabData;
+	}
+
+	/**
+	 * Freespace Flag features for slabs.
+	 */
+
+	@property
+	bool supportsFreeSpace() const {
+		return isAppendableSizeClass(sizeClass);
+	}
+
+	@property
+	ref shared(Bitmap!256) freeSpaceFlags() {
+		assert(isSlab(), "freeSpaceFlags accessed on non slab!");
+		assert(supportsFreeSpace, "freeSpaceFlags not supported!");
+
+		return _metadata.slabData.freeSpaceData.freeSpaceFlags;
+	}
+
+	bool setFreeSpace(uint index, size_t freeSpace) {
+		assert(isSlab(), "setFreeSpace accessed on non slab!");
+		assert(freeSpace <= slotSize, "freeSpace exceeds alloc size!");
+		assert(index < slotCount, "index is out of range!");
+
+		if (!supportsFreeSpace) {
+			return false;
+		}
+
+		if (freeSpace == 0) {
+			freeSpaceFlags.setBitValueAtomic!false(index);
+			return true;
+		}
+
+		// Encode freespace and write it to the last byte (or two bytes) of alloc.
+		writePackedFreeSpace(cast(ushort*) (address + slotSize - 2),
+		                     freeSpace & ushort.max);
+		freeSpaceFlags.setBitValueAtomic!true(index);
+
+		return true;
+	}
+
+	size_t getFreeSpace(uint index) {
+		assert(isSlab(), "getFreeSpace accessed on non slab!");
+		assert(index < slotCount, "index is out of range!");
+
+		if (!(supportsFreeSpace && slabData.valueAt(index)
```

Comment:
Is getting the free space on a slot that is not allocated valid behavior? And if not, then why isn't it a precondition? And anyway, there is no synchronization on that field. You are jumping through hoops to ensure this is not a problem for the free space flag, so I'm puzzled as to how this was missed.

Reply:
I think you are raising a very interesting point that I did not think of before, but, unfortunately, this is not an appropriate solution.

A pointer, supplied by the user is clearly an indication of an error in user's code. We don't have to handle these, after all, the user could go snoop anywhere in the memory and change the allocator's data structures, then what? While we need to account for pointer coming from the stack, data segments, tls, and whatever else, we shouldn't expect to see ponter comming from an unallocated slot in a valid program... Except in one specific instance: 0 length range can bleed into the next slot.

Consider:
```d
ubyte[] array = new ubyte[16]; // An appendable size class.
getCapacity(array[16 .. 16]); // The base pointer is in in the next slot.
```

This case needs to be handled properly. This solution ain't cutting it: the slot might be allocated, and returning anything would still be invalid! Maybe the solution here is to consider that empty slices always have 0 capacity.